crates/resolver-tests/src/lib.rs

```diff
@@ -298,110 +466,151 @@ impl SatResolver {
         // But things run faster if we let it spend some time figuring out how the constraints interact before we add assumptions.
         solver
             .solve()
             .expect("docs say it can't error in default config");
 
         SatResolver(Rc::new(RefCell::new(SatResolverInner {
             solver,
+            old_root_vars: Vec::new(),
             var_for_is_packages_used,
+            var_for_is_packages_features_used,
             by_name,
         })))
     }
 
-    pub fn sat_resolve(&self, root_summary: &Summary, _root_cli_features: &CliFeatures) -> bool {
-        let mut s = self.0.borrow_mut();
+    pub fn sat_resolve(&self, root_summary: &Summary, root_cli_features: &CliFeatures) -> bool {
+        let SatResolverInner {
+            solver,
+            old_root_vars,
+            var_for_is_packages_used,
+            var_for_is_packages_features_used,
+            by_name,
+        } = &mut *self.0.borrow_mut();
+
         let mut assumption = vec![];
-        let mut this_call = None;
-
-        // the starting `deps` need to be satisfied
-        for dep in root_summary.dependencies().iter() {
-            let empty_vec = vec![];
-            let matches: Vec<varisat::Lit> = s
-                .by_name
-                .get(dep.package_name().as_str())
-                .unwrap_or(&empty_vec)
-                .iter()
-                .filter(|&p| dep.matches_id(*p))
-                .map(|p| s.var_for_is_packages_used[p].positive())
-                .collect();
-            if matches.is_empty() {
-                return false;
-            } else if matches.len() == 1 {
-                assumption.extend_from_slice(&matches)
-            } else {
-                if this_call.is_none() {
-                    let new_var = s.solver.new_var();
-                    this_call = Some(new_var);
-                    assumption.push(new_var.positive());
-                }
-                let mut matches = matches;
-                matches.push(this_call.unwrap().negative());
-                s.solver.add_clause(&matches);
+
+        let root_pkg = root_summary.package_id();
+
+        // root package is always used
+        let root_var = solver.new_var();
+        assumption.push(root_var.positive());
+
+        // define new vars for each root feature
+        let var_for_root_features_used: HashMap<InternedString, varisat::Var> = root_summary
+            .features()
+            .keys()
+            .map(|&f| (f, solver.new_var()))
+            .collect();
+
+        for (_, root_feature_var) in &var_for_root_features_used {
+            solver.add_clause(&[root_feature_var.negative(), root_var.positive()]);
+        }
+
+        // replace previous root package
+        var_for_is_packages_used.insert(root_pkg, root_var);
+        var_for_is_packages_features_used.insert(root_pkg, var_for_root_features_used);
```

Comment:
the root package variables are inserted in these maps to allow reusing `add_pkg_clauses()` and `process_feature_value()` functions.